text/3759-cargo-supported-targets.md

```diff
@@ -0,0 +1,592 @@
+`supported-targets`. The rationale is that an example, test, or benchmark has access to the
+package's library and binaries, and so it must respect the `supported-targets` of the package.
+
+### Compatibility of `[build-dependencies]`
+
+What makes `[build-dependencies]` unique is that they are built for the host computer, and not the
+selected target. As such, they are not restrained by the `supported-targets` of the package. Hence,
+all dependencies are allowed in the `[build-dependencies]` table. However, a build error could be
+raised if one of the build dependencies does not support the _host-tuple_ at build time.
+
+A problem can arise if a crate's build script depends on a package that does not support `target_os
+= "windows"` for example. It would be possible to only allow dependencies supporting all targets in
+`[build-dependencies]`.
+
+
+### Platform-specific dependencies
+
+Platform-specific dependencies are dependencies under the `[target.**]` table. This includes normal
+dependencies, build-dependencies, and dev-dependencies. Rules could be defined to ensure that
+platform-specific dependencies are declared correctly.
+
+When platform-specific dependencies are declared, the conditions under which they are declared
+should be a subset of each dependency's `supported-targets`. For example, a dependency declared
+under `[target.'cfg(target_os = "linux")'.dependencies]` should at least support the `linux` OS.
+
+For regular dependencies and dev-dependencies, it would suffice for a platform-specific dependency
+to support the _intersection_ of the package's supported-targets, and the target conditions it is
+declared under. For example:
+```toml
+[package]
+# ...
+supported-targets = 'cfg(target_os = "linux")'
+
+[target.'cfg(target_pointer_width = "64")'.dependencies]
+foo = "0.1.0"
+```
+Here, it would suffice for `foo` to support `cfg(all(target_os = "linux", target_pointer_width =
+"64"))`.
+
+This would ensure that a package properly uses dependencies that are not available on all targets.
+Assuming that the crate `io-uring` has `supported-targets = 'cfg(target_os = "linux")'`, a crate
+could depend on it using:
+```toml
+[package]
+# ...
+
+[target.'cfg(target_os = "linux")'.dependencies]
+io-uring = "0.1.0"
+```
+This would not be required if the package itself had `supported-targets = 'cfg(target_os =
+"linux")'`, or an even stricter set.
+
+### Artifact dependencies
+
+If an artifact dependency has a `target` field, then the dependency would not be checked against the
+package's `supported-targets`. However, the selected `target` for the dependency would need to be
+compatible with the dependency's `supported-targets`, or else an error is raised. If the artifact
+dependency does not have a `target` field, then it would be checked against the package's
+`supported-targets`, like any other dependency.
+
+
+## Eliminating unused dependencies from `Cargo.lock`
```

Comment:
The main focus of this is limiting inclusion in `cargo vendor` so that should be clearly called out here

Reply:
Ref: https://github.com/rust-lang/cargo/issues/7058